src/integrator/cuda/cuda_weights.cu

```diff
@@ -360,12 +386,184 @@ __global__ void modify_weights_ssf_kernel_1d(
+      if (threadIdx.x == 0)
+        weights_device[ipt] = 0.;
+      __syncwarp();
+      continue;
+    }
+
+    // Initialize each counter to 0
+    if (threadIdx.x == 0) {
+      jCounter[0] = 0;
+    }
+    __syncwarp();
+
+    // Each thread will process an iCenter. Atomic operations are used to assign
+    // an iCenter value to each thread.
+    int iCenter = atomicAdd(jCounter, 1);
+    if (iCenter >= iParent) iCenter++; // iCenter == iParent is skipped
+
+    // The entire warp processes the same jCenter value at the same time
+    int jCenter = 0;
+
+    const double* local_rab = RAB + iCenter * LDatoms;
+    double ri = local_dist_scratch[ iCenter ];
+    double ps = 1.;
+    int iCount = 0; 
+    int cont = (iCenter < natoms);
+
+    // We will continue iterating until all of the threads have cont set to 0
+    while (__any_sync(0xffffffff, cont)) {
+      if (cont) {
+        double2 rj[weight_unroll/2];
+        double2 rab_val[weight_unroll/2];
+        double mu[weight_unroll];
+        iCount += weight_unroll;
+
+        #pragma unroll
+        for (int k = 0; k < weight_unroll/2; k++) {
+          rj[k]      = *((double2*)(local_dist_scratch + jCenter) + k);
+          rab_val[k] = *((double2*)(local_rab          + jCenter) + k); 
+        }
+
+        #pragma unroll
+        for (int k = 0; k < weight_unroll/2; k++) {
+          mu[2*k+0] = (ri - rj[k].x) * rab_val[k].x; // XXX: RAB is symmetric
+          mu[2*k+1] = (ri - rj[k].y) * rab_val[k].y; 
+        }
+
+        #pragma unroll
+        for (int k = 0; k < weight_unroll; k++) {
+          if((iCenter != jCenter + k) && (jCenter + k < natoms)) {
+            mu[k] = sFrisch( mu[k] );
+            ps *= mu[k];
+          }
+        }
+
+        // A thread is done with a iCenter based on 2 conditions. Weight tolerance
+        // Or if it has seen all of the jCenters
+        if( !(ps > weight_tol && iCount < LDatoms )) {
+          // In the case were the thread is done, it begins processing another iCenter
+          sum += ps;
+          iCenter = atomicAdd(jCounter, 1);
+          if (iCenter >= iParent) iCenter++;
+
+          // If there are no more iCenters left to process, it signals it is ready to exit
+          cont = (iCenter < natoms);
+          ri = local_dist_scratch[ iCenter ];
+          local_rab = RAB + iCenter * LDatoms;
+          ps = 1.;
+          iCount = 0;
+        }
+      }
+      // Wraps jCenter around. This was faster than modulo
+      jCenter += weight_unroll;
+      jCenter = (jCenter < LDatoms) ? jCenter : 0;
+    }
+
+    // All of the threads then sum their contributions. Only thread 0 needs to add the parent
+    // contribution.
+    __syncwarp();
+    sum = warpReduceSum(sum);
+    if (threadIdx.x == 0) {
+      sum += parent_weight;
+      weights_device[ipt] *= parent_weight / sum;
+    }
+
+    __syncwarp();
+
+  }
+}
+
 
+void cuda_reciprocal(size_t length, double* vec, cudaStream_t stream) {
```

Comment:
I would add this into one of the cuda utility files: not really specific to the weights (though it is the only place that its used)

Reply:
Is there a .cu utility file to put the kernel itself in? Or is it okay if that stays in the weights file? If not, I can make one

Reply:
I would probably move this to `xc_cuda_util`, but we may need to do some refactoring in a bit anyways, so we can probably hold off for a bit on this